src/components/menu/menu-item/menu-item.component.js

```diff
@@ -1,33 +1,152 @@
+import React, {
+  useRef, useState, useEffect
+} from 'react';
 import PropTypes from 'prop-types';
 import classNames from 'classnames';
 import StyledMenuItemWrapper from './menu-item.style';
 import { StyledSubmenu, StyledSubmenuItem, StyledSubmenuTitle } from '../submenu-block/submenu.style';
 import OptionHelper from '../../../utils/helpers/options-helper';
 import Link from '../../link';
+import Events from '../../../utils/helpers/events';
 
-const MenuItem = ({
+const MenuItem = React.forwardRef(({
   submenu,
   children,
   href,
   to,
   menuType,
   onClick,
   target,
-  submenuDirection,
+  submenuDirection = 'right',
   icon,
   selected,
-  routerLink
-}) => {
+  routerLink,
+  isFirstElement,
+  handleKeyDown
+}, ref) => {
+  const [isOpen, setIsOpen] = useState(false);
+  const [openByArrowUp, setOpenByArrowUp] = useState(false);
+  const submenuItemsRefs = submenu && useRef(React.Children.map(children, child => child.ref || React.createRef()));
+  const ifRef = () => !submenu && { ref };
+  let actualIndex;
+  const setFocusToSubmenuElement = (event, index) => {
+    if (event) {
+      event.preventDefault();
+    }
+
+    const findIndex = () => {
+      if (submenuItemsRefs.current[index] && submenuItemsRefs.current[index].current.tagName === 'DIV') {
+        if (Events.isDownKey(event)) {
+          return index + 1;
+        }
+        return index - 1;
+      }
+      return index;
+    };
+    const tempIndex = findIndex();
+    let newIndex = tempIndex;
+    actualIndex = tempIndex;
+
+    if (index === submenuItemsRefs.current.length) {
+      newIndex = 0;
+    } else if (index === -1) {
+      newIndex = submenuItemsRefs.current.length - 1;
+    }
+
+    submenuItemsRefs.current[newIndex].current.focus();
+  };
+
+  useEffect(() => {
+    if (submenu && isOpen) {
+      if (!openByArrowUp) {
+        submenuItemsRefs.current[0].current.focus();
+      } else {
+        submenuItemsRefs.current[submenuItemsRefs.current.length - 1].current.focus();
+      }
+    }
+  }, [isOpen, openByArrowUp, submenu, submenuItemsRefs]);
+
+  const onCloseSubmenu = () => {
+    setIsOpen(false);
+    setOpenByArrowUp(false);
+  };
+
+  const onKeyDownSubmenu = (ev, index) => {
+    if (!isOpen) {
+      if (Events.isEnterKey(ev) || Events.isSpaceKey(ev) || Events.isDownKey(ev)) {
+        setIsOpen(true);
+      }
+      if (Events.isUpKey(ev)) {
+        setOpenByArrowUp(true);
+        setIsOpen(true);
+      }
+    }
+
+    if (isOpen) {
+      if (Events.isDownKey(ev)) {
+        setFocusToSubmenuElement(ev, index + 1);
+      }
+      if (Events.isUpKey(ev)) {
+        setFocusToSubmenuElement(ev, index - 1);
+      }
+      if (
+        Events.isLeftKey(ev)
+        || Events.isRightKey(ev)
+        || (Events.isShiftKey(ev) && Events.isTabKey(ev))
+        || Events.isTabKey(ev)
+      ) {
+        onCloseSubmenu();
+      }
+      if (Events.isHomeKey(ev)) {
+        submenuItemsRefs.current[0].current.focus();
```

Comment:
We need to ensure for every keypress that we handle that we `e.preventDefault()`. On a submenu, we might also need `e. stopPropagation()` to prevent the `Menu` key handler from processing the event too.